### ISSUE TITLE
SCSt Binder Doesn't Seem to support concurrency Consumer Property

### ISSUE THREAD

**Issue by Mrc0113** [SolaceProducts/solace-spring-cloud#7]
Unless I'm overlooking something it seems that the Solace Spring Cloud Stream Binder does not support the `concurrency` [Consumer Property](https://cloud.spring.io/spring-cloud-static/spring-cloud-stream/current/reference/html/spring-cloud-stream.html#_consumer_properties). If I set this value to 10 I would expect to see 10 consumers bound to the generated queue however I always only see 1. 

Example config file: 
```
spring:
  cloud:
    stream:
      function:
        definition: process
      bindings:
        process-in-0:
          destination: myQueue
          group: foo
          consumer:
            concurrency: 10
      solace:
        bindings:
          process-in-0:
            consumer:
              queueAdditionalSubscriptions: 'pub/sub/plus/>'
solace:
  java:
    host: localhost:55555
    msgVpn: default
    clientUsername: default
    clientPassword: default
```

**Comment by Mrc0113** [comment]
Solace Community Discussion around this as well: https://solace.community/discussion/284/concurrency-property-with-solace-spring-cloud-stream-api#latest

**Comment by Nephery** [comment]
Yeah, our binder is missing support for this property.

I think the fix for this is probably to duplicate the flow consumer inside `JCSMPInboundChannelAdapter` to scale up to the number specified by that property.

Also, as a preventative measure, we should throw a `ProvisioningException` at startup time (this will fail the app from starting) if `concurrency` is greater than `1` AND the consumer group's queue's access type is set to `exclusive`.

I don't think concurrency really applies to polled consumers either, so we should do one of the following if `concurrency` is greater than `1` for polled consumers:
- Ideally, fail at provisioning to prevent the app from starting
- Otherwise, fail the flow consumer from starting (but this wouldn't prevent the app from starting).

**Comment by Nephery** [comment]
> Ideally, fail at provisioning to prevent the app from starting

I'm starting to have second thoughts on this suggestion for dealing with polled consumers. We probably should just do the second suggestion since its related to consumers and not provisioning itself.

**Comment by Mrc0113** [comment]
> I think the fix for this is probably to duplicate the flow consumer inside JCSMPInboundChannelAdapter to scale up to the number specified by that property.
Yes, this makes sense to me. 

> Also, as a preventative measure, we should throw a ProvisioningException in the SolaceQueueProvisioner (this will fail the app from starting) if concurrency is greater than 1 AND the consumer group's queue's access type is set to exclusive.
Hmm. I don't think we want to stop the app from starting b/c what if they have 2 consuming functions, one that is exclusive and one that is non-exclusive? Maybe we just make a log entry that says the concurrency option is not supported for exclusive consumers and leave it at 1 for those?   

> I don't think concurrency really applies to polled consumers either, so we should do one of the following if concurrency is greater than 1 for polled consumers:
Same comment as above. I haven't used "polled consumers" but if they can be combined with a message-driven one then we don't want to fail. I briefly searched through the rabbit binder code and I didn't notice them taking `concurrency` into account w/ polled consumers but I could have missed something. 

Something else to consider, it looks like when using "reactive" that the framework itself says the `concurrency` parameter is not supported [here](https://github.com/spring-cloud/spring-cloud-stream/blob/10fc8dea30608b34266323fbe2831dfed7b4a29b/spring-cloud-stream/src/main/java/org/springframework/cloud/stream/function/FunctionConfiguration.java#L369)

**Comment by Nephery** [comment]
> Hmm. I don't think we want to stop the app from starting b/c what if they have 2 consuming functions, one that is exclusive and one that is non-exclusive?

You have to define `concurrency` for each binding, so I'm not sure what the concern is for this. In a way, you could see this as being similar to how we fail startup if someone were to try to configure their binding to reconnect to an existing queue with a mismatched access type.

Honestly, I'm fine with either solution; Whether we fail app at startup or log a warning and ignoring the `concurrency` property (which is basically what you're suggested).

> Something else to consider, it looks like when using "reactive" that the framework itself says the concurrency parameter is not supported here

Quoting the comment in https://github.com/olegz/spring-cloud-stream/commit/8bba5746903f092b82f5a96b39f901bb9520b683 :

> Providing that project reactor maintains it's own concurrency mechanisms, using stream's concurrency setting can lead to unexpected results including loss of messages.
This fixes it by disallowing concurrency setting when reactive consumers are used.

[This commit is currently slated for 3.1.0](https://github.com/spring-cloud/spring-cloud-stream/issues/1977). So while it is currently "possible" to use concurrency with reactive consumers, I don't think its worth the effort for us to do anything if support for this is going to be removed in the near future. And I don't think we could've done anything anyway since this looks to be outside of the binder's jurisdiction (i.e. the binder doesn't know whether its a imperative or reactive consumer).

**Comment by Mrc0113** [comment]
| You have to define concurrency for each binding, so I'm not sure what the concern is for this.

Good point, agree no concern now. I'm fine with either option as well. Log or Fail. 

| This commit is currently slated for 3.1.0. So while it is currently "possible" to use concurrency with reactive consumers, I don't think its worth the effort for us to do anything if support for this is going to be removed in the near future. And I don't think we could've done anything anyway since this looks to be outside of the binder's jurisdiction (i.e. the binder isn't able to tell the difference between imperative and reactive consumers).

Agree, looks like they'll handle at framework level

**Comment by Nephery** [comment]
@Mrc0113 and I talked more offline.

### Potential Solutions

We have 3 potential solutions if the queue is exclusive and concurrency > 1:
1. Fail before we provision any queues, preventing the app from running
1. Fail the creation of that binding's consumer, which will not prevent the app from running, but the failed consumer would be missing.
1. Ignore concurrency and log a warning that we're treating it as concurrency=1

* Solution 1 is a fail-fast and makes it pretty clear to you that something is wrong/mismatched. This should be fine since non-exclusive queues are used by default with consumer-groups. And if we properly document that concurrency should not be used together with an exclusive queue, then it should be fine. Of course with a descriptive error message describing the problem.
* Solution 2 was the least ideal of the 3 since it intentionally creates a failure that could easily be overlooked.
* Solution 3 is also OK from a technical standpoint, but then it introduces the problem that the config would no longer fully reflect the setup. While we do have situations where this would occur (e.g. when `provisionDurableQueue` is `false`), we shouldn't intentionally introduce more if we can avoid it.

**TLDR:** We chose option 1

### Concurrency With Bindings That Use Anonymous-Groups

Anonymous groups use temporary queues. Temporary queues are strictly exclusive and do not support multiple consumers. For now, our only solution is to reject concurrency > 1 on anonymous-queues. This should be a hard failure as well.

### Summary

All that to say, the solution for this should be:
* Duplicate the flow consumer inside `JCSMPInboundChannelAdapter` to scale up to the number specified by `concurrency`.
* Fail at provisioning time any of the following are true:
  * concurrency > 1 and queue is exclusive
  * concurrency > 1 and binding is associated to an anon-group
* Document these limitations in the `concurrency` property's description in the `README.md`

**Comment by Nephery** [comment]
So I misunderstood how the binder handles failures. I thought that if the queue provisioner failed, it would shutdown the app. But that is not the case; when the binder receives a provisioning error, it periodically (~30 sec) retries to create that binding.
This means that we should replace solutions 1 and 2 with:

1. Fail the provisioning of the queue, SCSt will attempt to recreate the binding after some interval
2. Fail the provisioning of the consumer (this happens after the binding's queue is provisioned), SCSt will attempt to recreate the binding after some interval

They're basically the same as before, except that the app will start in both scenarios and SCSt will periodically retry to create the failed bindings.

In the end, we'll be going with the new solution 1 since we don't want to create durable queues for no reason.

**Comment by FWinkler79** [comment]
Hi there, 

I read through the discussion, since I am facing the exact same problem: concurrency is not supported.
I really hope this is coming soon, as it is currently a blocking issue for us.

Above I read:
> You have to define concurrency for each binding.

Note, that there is also a `spring.cloud.stream.default.consumer` setting where concurrency can be set. Unless a binding explicitly overrides that value, the default will be taken. Not sure this is relevant for your discussion, though.

Any idea, when concurrency (for non-exclusive access) will be supported?

**Comment by Nephery** [comment]
@FWinkler79 

> Note, that there is also a spring.cloud.stream.default.consumer setting where concurrency can be set. Unless a binding explicitly overrides that value, the default will be taken. Not sure this is relevant for your discussion, though.

The `concurrency` config option is owned by the core framework itself, so setting a default `concurrency` should just work without us needing to write any special code in the binder. What I meant in that comment was that the `concurrency` config option is effectively applied on a per-binding basis.

> Any idea, when concurrency (for non-exclusive access) will be supported?

In the next 1.1.0 release which should be released in the coming weeks.
